main/src/library/Process.flix

```diff
@@ -17,34 +17,77 @@
 ///
 /// An effect used to start a process outside the JVM.
 ///
+/// All operations on this effect are infallible.
+/// If an operation fails the handler must deal with it.
+///
 eff Process {
 
     ///
-    /// Immediately executes the command `cmd` passing the arguments `args`.
+    /// Immediately executes the command `cmd` with the arguments `args`, by the path `cwd` and with the environmental `env`
     ///
-    def exec(cmd: String, args: List[String]): Unit
-
+    def execWithCwdAndEnv(cmd: String, args: List[String], cwd: Option[String], env: Map[String, String]): ProcessHandle
 }
 
 mod Process {
 
+    use IoError.ErrorKind
+    use IoError.IoError
+
+    import java.lang.IllegalArgumentException
+    import java.lang.SecurityException
+    import java.io.IOException
     import java.lang.ProcessBuilder
     import java.lang.Runtime
+    import java.io.{File => JFile}
+
+    ///
+    /// Executes the command `cmd` with the arguments `args`
+    ///
+    pub def exec(cmd: String, args: List[String]): ProcessHandle \ Process =
+        Process.execWithCwdAndEnv(cmd, args, None, Map.empty())
+
+    ///
+    /// Executes the command `cmd` with the arguments `args`, by the path `cwd`
+    ///
+    pub def execWithCwd(cmd: String, args: List[String], cwd: Option[String]): ProcessHandle \ Process =
+        Process.execWithCwdAndEnv(cmd, args, cwd, Map.empty())
+
+    ///
+    /// Executes the command `cmd` with the arguments `args` and with the environmental `env`
+    ///
+    pub def execWithEnv(cmd: String, args: List[String], env: Map[String, String]): ProcessHandle \ Process =
+        Process.execWithCwdAndEnv(cmd, args, None, env)
 
     ///
     /// Handles the `Process` effect of the given function `f`.
     ///
     /// In other words, re-interprets the `Process` effect using the `Exec` effect.
     ///
-    pub def handle(f: a -> b \ ef): a -> b \ (ef - Process) + {Exec, IO} = x ->
+    pub def handle(f: a -> b \ ef): a -> Result[IoError, b] \ (ef - Process) + {Exec, IO} = x ->
         run {
-            f(x)
+            Ok(f(x))
         } with Process {
-            def exec(cmd, args, k) = region rc {
-                let arr = List.toArray(rc, cmd :: args);
-                let pb = new ProcessBuilder(arr);
-                pb.start();
-                k()
+            def execWithCwdAndEnv(cmd, args, cwd, env, handler) = region rc {
+                try {
+                    let arr = List.toArray(rc, cmd :: args);
+                    let pb = new ProcessBuilder(arr);
+                    foreach(e <- Map.toList(env)) {
+                        pb.environment().put(fst(e), snd(e));
+                        ()
+                    };
+                    match cwd {
+                        case Some("") => throw new IllegalArgumentException("given an empty directory as an argument")
```

Comment:
I think we should just drop this case. Its not the just the empty string that is illegal right?

So either you have to check that it is a real path or we let Java deal with it and we catch the appropriate exception and translate it to Result.Err.